### PR TITLE
WIP: Mysql: Fix Model.count returning NaN. Closes #3019

### LINKED ISSUE #3019: Model.count returning NaN because Mysql returns an empty string
### Problem

`Model.count` is returning `NaN` because Mysql returns an empty string when grouping.

```MySQL
SELECT count(*) AS `count` FROM `Model` AS `Model` WHERE (`Model`.`deleted_at` IS NULL AND `Model`.`a_model`='Post' AND `Model`.`a_foreign_key`=186);
# count: 0
```

```MySQL
SELECT count(*) AS `count` FROM `Model` AS `Model` WHERE (`Model`.`deleted_at` IS NULL AND `Model`.`a_model`='Post' AND `Model`.`a_foreign_key`=186) GROUP BY `Model`.`user_id`;
# count: ''
```

If there are values to group up, everything works as expected, but if not, MySQL returns an empty string instead of a `0` (zero).

### Solution

Return parsed Int, or `0` at [queryInterface](https://github.com/sequelize/sequelize/blob/master/lib/query-interface.js#L734);

### PR DESCRIPTION
**Pull request by BridgeAR**
This will fix DOUBLE and REAL being returned as numbers instead of strings. This will also fix the issue with iojs 2.0 and the missing docs entry that .count supports paranoid. 
I also updated some internal dependencies and changed some to get patch updates instead of being fixed.

### CODE REVIEW

**Comment by BridgeAR** [comment]
This might fix #3019. I'm not really sure though, since the query always returned as an empty array instead of a string as described in the issue.

File: test/integration/dialects/mysql/connector-manager.test.js [hunk 1]
```diff
@@ -15,19 +15,23 @@ if (Support.dialectIsMySQL()) {
         , self = this;
 
       return User.sync({force: true}).then(function() {
-        return User.create({username: 'user1'}).then(function() {
+        return User.count({group: ['id']}).then(function(count) {
+          expect(count.length).to.equal(0);
```
Comment (cusspvz) [inline test/integration/dialects/mysql/connector-manager.test.js]: shouldn't be `expect(count).to.equal(0)`?

File: test/integration/dialects/mysql/connector-manager.test.js [hunk 2]
```diff
@@ -15,19 +15,23 @@ if (Support.dialectIsMySQL()) {
...
-              return User.count().then(function(count) {
-                expect(count).to.equal(1);
-                spy();
-                if (!spy.calledTwice) {
-                  throw new Error('Spy was not called twice');
-                }
-              });
-            });
+          });
+        }).then(function() {
```
Comment (cusspvz) [inline test/integration/dialects/mysql/connector-manager.test.js]:
```
})
.delay(1000)
.then(function () {
```

File: lib/query-interface.js [hunk 3]
```diff
@@ -718,17 +718,20 @@ module.exports = (function() {
...
+        if (
+          dataType instanceof DataTypes.DECIMAL ||
+          dataType instanceof DataTypes.INTEGER ||
+          dataType instanceof DataTypes.FLOAT ||
+          dataType instanceof DataTypes.BIGINT ||
+          dataType instanceof DataTypes.DOUBLE ||
+          dataType instanceof DataTypes.REAL
+        ) {
+          // Convert the result string to a number
+          result = +result;
```
Comment (cusspvz) [inline lib/query-interface.js]: `result = +result || 0;`

Comment (cusspvz) [inline lib/query-interface.js]: This will cover `NaN` cases.

File: test/integration/dialects/mysql/connector-manager.test.js (hunk 1 above)
Comment (BridgeAR) [inline test/integration/dialects/mysql/connector-manager.test.js]: Nop - it's an array that is returned

File: test/integration/dialects/mysql/connector-manager.test.js (hunk 2 above)
Comment (BridgeAR) [inline test/integration/dialects/mysql/connector-manager.test.js]: I prefer to write it with less indentation

File: lib/query-interface.js (hunk 3 above)
Comment (BridgeAR) [inline lib/query-interface.js]: +'' === 0

File: test/integration/dialects/mysql/connector-manager.test.js (hunk 1 above)
Comment (cusspvz) [inline test/integration/dialects/mysql/connector-manager.test.js]: Should be an integer.
[`count([options]) -> Promise.<Integer>](http://sequelize.readthedocs.org/en/latest/api/model/index.html?highlight=count#countoptions-promiseinteger)

**Comment by cusspvz** [comment]
> I'm not really sure though, since the query always returned as an empty array instead of a string as described in the issue.

The root of this problem is that *mysql* returns an empty string when `count` is 0, and as *sequelize* wasn't parsing numeric results, it was returning `NaN` (because somewhere on sequelize's or our code, integers were being parsed).

File: lib/query-interface.js (hunk 3 above)
Comment (cusspvz) [inline lib/query-interface.js]: Well, seems valid! :)
I though `+'' === NaN`, sorry.

File: test/integration/dialects/mysql/connector-manager.test.js (hunk 1 above)
Comment (BridgeAR) [inline test/integration/dialects/mysql/connector-manager.test.js]: Thx for pointing out that the docs are wrong

**Comment by BridgeAR** [comment]
If you use the group option, plain will be set. This has to be that way, since the return value might contain more than one entry. So if plain is set, the result will look like: [{count: 1}] or like [] if count is zero. The line you pointed at it will never be reached.
To solve your NaN it would be nice to get your code piece that fails, so I can help.

**Comment by cusspvz** [comment]
> To solve your NaN it would be nice to get your code piece that fails, so I can help.

I think this PR actually fixes it.

> So if plain is set, the result will look like: [{count: 1}] or like [] if count is zero. The line you pointed at it will never be reached.

Didn't knew about `group` option. Seems right!

I think this PR is ready to be merged.
Great work @BridgeAR !! Thanks :)

File: test/integration/dialects/mysql/connector-manager.test.js (hunk 1 above)
Comment (mickhansen) [inline test/integration/dialects/mysql/connector-manager.test.js]: It really should return a single value though, not an array.

File: lib/query-interface.js (hunk 3 above)
Comment (mickhansen) [inline lib/query-interface.js]: BIGINT's cant be expressed as javascript numbers, this is an unsafe change.

Comment (BridgeAR) [inline lib/query-interface.js]: Well I did not change that ;) 
I'll remove the BIGINT

File: test/integration/dialects/mysql/connector-manager.test.js (hunk 1 above)
Comment (BridgeAR) [inline test/integration/dialects/mysql/connector-manager.test.js]: It's a plain result with an array with one entry. The entry contains an object with count plus the attributes you provided as option.

Comment (mickhansen) [inline test/integration/dialects/mysql/connector-manager.test.js]: That sounds like a regression then. It should result in a primitive afaik.

**Comment by BridgeAR** [comment]
After thinking about it: The only function call to convert the number is coming from the `aggregate` function. There could be big numbers parsed, but this is not very likely. This would be a BC if we change it now.

File: test/integration/dialects/mysql/connector-manager.test.js [hunk 4]
```diff
@@ -15,19 +15,23 @@ if (Support.dialectIsMySQL()) {
         , self = this;
```
Comment (janmeier) [inline test/integration/dialects/mysql/connector-manager.test.js]: Connector-manager.test might be just about the worst place to add this test :D

File: test/integration/dialects/mysql/connector-manager.test.js (hunk 1 above)
Comment (janmeier) [inline test/integration/dialects/mysql/connector-manager.test.js]: This is probably related to the issue mentioned on IRC the other day @mickhansen 

When grouping count will return several rows - we haven't really decided what should happen it this case

**Pull request closed without merging**